### PR TITLE
Add Sign in with Apple Display Name API and unit test

### PR DESCRIPTION
**Pull request by aiwenisevan**
### Discussion

  * This resolves [Sign in with Apple display name not fulfilled [235848270] - Buganizer](http://b/235848270)

### Testing

  * Unit testing Passed for both `FIROAuthProviderTest` and `FIRAuthTest`

### API Changes

  * Internal API proposal: [Sign in with Apple Display Name API](http://doc/17BPHhMVfSfmKmELjt0m2F1x7YrM2TQUO13sgmFn5MSM#heading=h.x94y9zzc13xn)

### CODE REVIEW

File: FirebaseAuth/Sources/AuthProvider/OAuth/FIROAuthCredential_Internal.h
Comment (rosalyntan) [inline FirebaseAuth/Sources/AuthProvider/OAuth/FIROAuthCredential_Internal.h:67]: Instead of defining a new init method, can you just add a `displayName` parameter to the existing method above, and then update the existing callers of that method to pass in a `nil` `displayName`?

File: FirebaseAuth/Sources/Backend/RPC/FIRVerifyAssertionRequest.h
Comment (rosalyntan) [inline FirebaseAuth/Sources/Backend/RPC/FIRVerifyAssertionRequest.h:102]: You'll need to make changes in `FIRVerifyAssertionRequest.m` as well.

Comment (aiwenisevan) [reply FirebaseAuth/Sources/Backend/RPC/FIRVerifyAssertionRequest.h:102]: Thanks, Have made changes

File: FirebaseAuth/Tests/Unit/FIRAuthTests.m
Comment (rosalyntan) [inline FirebaseAuth/Tests/Unit/FIRAuthTests.m:1428]: nit: Can you name this `testSignInWithOAuthCredentialWithDisplayNameSuccess` to make it more clear the purpose of this test?

File: FirebaseAuth/Tests/Sample/Sample/GoogleAuthProvider.m
Comment (aiwenisevan) [reply FirebaseAuth/Tests/Sample/Sample/GoogleAuthProvider.m:25]: merged

File: FirebaseAuth/Tests/Unit/FIRAuthTests.m
Comment (aiwenisevan) [reply FirebaseAuth/Tests/Unit/FIRAuthTests.m:1435]: Do I need to clean the other ones and only test displayName? Or I can just add displayName to test?

**Comment by Ricky-Spanish** [comment]
This should be merged as soon as possible since apps are currently being rejected from the App Store due to the issue. MR looks good to me 👍

**Review by Ricky-Spanish — APPROVED** [review]
Looks good, approved. 👍

**Comment by hamishjohnson** [comment]
Any idea if this will be released before the new year?

**Comment by nmunoz-mybudgeat** [comment]
Hi, this pull request has been in draft for quite some time. Any ETA on when it will be merged into master? Launching our app to production but this is really blocking us

**Comment by aiwenisevan** [comment]
> Any idea if this will be released before the new year - and will this also fix the email not appearing correctly? [firebase/flutterfire#9662](https://github.com/firebase/flutterfire/issues/9662)

Hi, this is still in progress and waiting for the backend part to be done. Do you mind describing the email appearing problem? So i can check it for you?

**Comment by nmunoz-mybudgeat** [comment]
> > Any idea if this will be released before the new year - and will this also fix the email not appearing correctly? [firebase/flutterfire#9662](https://github.com/firebase/flutterfire/issues/9662)
> 
> Hi, this is still in progress and waiting for the backend part to be done. Do you mind describing the email appearing problem? So i can check it for you?

would you mind sharing the backend Github issue so I can keep track of it?

**Comment by aiwenisevan** [comment]
> Hi, this pull request has been in draft for quite some time. Any ETA on when it will be merged into master? Launching our app to production but this is really blocking us

Hi, currently waiting for the backend to be ready, it maybe ready for next step at mid January.

**Comment by paulb777** [comment]
And there is no backend GitHub issue, so this issue is the best place to track publicly.

**Comment by szaboagoston** [comment]
Any update on this?

**Comment by brandellcassio** [comment]
Guys, any updates on this? Without this, a **lot** of Apps based on firebase are being rejected. You were working on this for three months, and it seems that the PR is not a big issue. So, it would be great to have this PR approved ASAP. 👍

**Comment by MahMoos** [comment]
If you encounter this issue using the [flutterfire-ui](https://pub.dev/packages/flutterfire_ui) package, you should know it has been deprecated. Everything worked fine for me in the newly created packages [firebase_ui_oauth](https://pub.dev/packages/firebase_ui_auth) and [firebase_ui_oauth_apple](https://pub.dev/packages/firebase_ui_oauth_apple) and I can retrieve Apple display name using it.

**Comment by kirilllapshev** [comment]
Would be amazing if you approve this PR. It is been blocking the launch of the new version of our app for a few weeks already! Thank you!

**Comment by ElixirMike** [comment]
The ticket mentions displayName not appearing, but this also includes email, correct? I'm receiving neither when signing in via apple on IOS and in my use-case, email is critical.

**Comment by Flucadetena** [comment]
Any EAT on when will this be merged? It is really a hustle to get and app published in the AppStore with this issue.

Thanks for the hard work.

**Comment by peterfriese** [comment]
> Any ETA on when will this be merged? It is really a hustle to get and app published in the AppStore with this issue.

Sorry for the hassle - we're waiting for some backend work to land.

Did you check out the code in [our Firebase Fundamentals video](https://youtu.be/HyiNbqLOCQ8)? This should help in the meantime.

**Comment by Flucadetena** [comment]
@peterfriese Thanks, just wanted to know how the fix is going. I'm working on a Flutter project and it is affected by this. So I'll wait for the fix.
Thanks anyway, hope it is solved soon.

**Comment by Ricky-Spanish** [comment]
Please merge this MR, many apps are being rejected from the App Store due to the issue.

**Comment by rosalyntan** [comment]
Hi folks, to give an update -- as @peterfriese mentioned, this PR is waiting for some backend changes to land (without those changes, this PR alone will not address your issue). The backend work is in progress, and once it's in we'll merge the client-side fix as well.

File: FirebaseAuth/Sources/AuthProvider/OAuth/FIROAuthCredential_Internal.h
Comment (peterfriese) [inline FirebaseAuth/Sources/AuthProvider/OAuth/FIROAuthCredential_Internal.h:56]: I think _displayName_ should be _full name_  in the description, like everywhere else.

**Review by peterfriese — CHANGES_REQUESTED** [review]

File: FirebaseAuth/Sources/AuthProvider/OAuth/FIROAuthCredential_Internal.h
Comment (rosalyntan) [reply FirebaseAuth/Sources/AuthProvider/OAuth/FIROAuthCredential_Internal.h]: Done.

File: FirebaseAuth/Sources/Backend/RPC/FIRVerifyAssertionRequest.h
Comment (rosalyntan) [reply FirebaseAuth/Sources/Backend/RPC/FIRVerifyAssertionRequest.h]: Done.

File: FirebaseAuth/Sources/Backend/RPC/FIRVerifyAssertionRequest.m
Comment (rosalyntan) [reply FirebaseAuth/Sources/Backend/RPC/FIRVerifyAssertionRequest.m:109]: The variable name (in line 112) is `kReturnIDPCredentialKey`. Since this isn't related to this PR and `IDP` is used in variable names elsewhere in the codebase, I'll leave this alone for consistency.

File: FirebaseAuth/Sources/AuthProvider/OAuth/FIROAuthProvider.m
Comment (rosalyntan) [reply FirebaseAuth/Sources/AuthProvider/OAuth/FIROAuthProvider.m:131]: Done (added in `FIROAuthProvder.h`)

File: FirebaseAuth/Sources/Public/FirebaseAuth/FIROAuthProvider.h
Comment (rosalyntan) [reply FirebaseAuth/Sources/Public/FirebaseAuth/FIROAuthProvider.h]: Resolved -- PR has been updated as per the updated API.

File: FirebaseAuth/Sources/Public/FirebaseAuth/FIROAuthProvider.h
Comment (paulb777) [inline FirebaseAuth/Sources/Public/FirebaseAuth/FIROAuthProvider.h:131]: Why does `fullName` need to be nullable?

Comment (rosalyntan) [reply FirebaseAuth/Sources/Public/FirebaseAuth/FIROAuthProvider.h:131]: So given that calling this method with a nil `fullName` is the same as calling `credentialWithProviderID` with the `apple.com` provider ID, `fullName` doesn't need to be nullable. However, given that this is an optional field, it seems more correct to make this nullable.

File: FirebaseAuth/Tests/Unit/FIRAuthTests.m
Comment (renkelvin) [inline FirebaseAuth/Tests/Unit/FIRAuthTests.m:1453]: Shall we add more checks on the user info?

**Review by renkelvin — APPROVED** [review]
LGTM, thanks!

**Review by peterfriese — APPROVED** [review]

File: FirebaseAuth/Tests/Unit/FIRAuthTests.m
Comment (rosalyntan) [reply FirebaseAuth/Tests/Unit/FIRAuthTests.m:1453]: For this test, I basically copied over the `testSignInWithCredentialSuccess` and added specific logic for verifying the display name. Since none of the logic besides the display name has changed, I removed some of the extra verification lines so it's easier to see exactly what this test is testing.

**Comment by byaruhaf** [comment]
So does this mean the display name is auto-updated so there is no need to update display name with appleIDCredential?

**Comment by peterfriese** [comment]
> So does this mean the display name is auto-updated so there is no need to update display name with appleIDCredential?

After receiving `ASAuthorization`  from Apple, call `appleCredential(withIDToken: rawNonce: fullName:)` to create the credential, and then sign in.

Here is the relevant code snippet from the [documentation](https://firebase.google.com/docs/auth/ios/apple):

```swift
@available(iOS 13.0, *)
extension MainViewController: ASAuthorizationControllerDelegate {

  func authorizationController(controller: ASAuthorizationController, didCompleteWithAuthorization authorization: ASAuthorization) {
    if let appleIDCredential = authorization.credential as? ASAuthorizationAppleIDCredential {
      guard let nonce = currentNonce else {
        fatalError("Invalid state: A login callback was received, but no login request was sent.")
      }
      guard let appleIDToken = appleIDCredential.identityToken else {
        print("Unable to fetch identity token")
        return
      }
      guard let idTokenString = String(data: appleIDToken, encoding: .utf8) else {
        print("Unable to serialize token string from data: \(appleIDToken.debugDescription)")
        return
      }
      // Initialize a Firebase credential, including the user's full name.
      let credential = OAuthProvider.appleCredential(withIDToken: idTokenString,
                                                        rawNonce: nonce,
                                                        fullName: appleIDCredential.fullName)
      // Sign in with Firebase.
      Auth.auth().signIn(with: credential) { (authResult, error) in
        if error {
          // Error. If error.code == .MissingOrInvalidNonce, make sure
          // you're sending the SHA256-hashed nonce as a hex string with
          // your request to Apple.
          print(error.localizedDescription)
          return
        }
        // User is signed in to Firebase with Apple.
        // ...
      }
    }
  }

  func authorizationController(controller: ASAuthorizationController, didCompleteWithError error: Error) {
    // Handle error.
    print("Sign in with Apple errored: \(error)")
  }

}
```

As you can see, you no longer need to update the user's `displayName` in a separate operation, it's all done for you.

**Comment by LilaQ** [comment]
@peterfriese I feel like I'm missing something. In the `Auth.auth().addStateDidChangeListener` the user still comes with a nil for a `displayName` for me

**Comment by peterfriese** [comment]
Hi LilaQ, I just verified this works as intended with Firebase 10.7.0.

If you're on the latest version of the SDK, can you share an [MVCE](https://stackoverflow.com/help/minimal-reproducible-example) with us? You can also check out the code in [this PR](https://github.com/FirebaseExtended/firebase-video-samples/pull/22) - it contains the updated call that updates the user's display name when signing in with Apple.

**Comment by jscottarmstrong** [comment]
@LilaQ are you using the local emulator? If I run my app connected to the local emulator for authentication, I don't see full name getting populated, but if I connect to the my live Firebase app, it is working as expected with 10.7.0.

**Pull request merged**
